src/join/composer.jl

```diff
@@ -182,11 +182,56 @@ function _propagate_join_metadata!(joiner::DataFrameJoiner, dfr_noon::AbstractDa
     return nothing
 end
 
+function _count_sortperm(input::Vector{Int})
+    isempty(input) && return UInt32[]
+    vmin, vmax = extrema(input)
+    delta = vmin - 1
+    Tc = vmax - delta < typemax(UInt32) ? UInt32 : Int
+    Tp = length(input) < typemax(UInt32) ? UInt32 : Int
+    _count_sortperm(input, zeros(Tc, vmax - delta + 1),
+                    Vector{Tp}(undef, length(input)), delta)
+end
+
+function _count_sortperm(input::Vector{Int}, count::Vector,
```

Comment:
Could you add a comment to explain what this function does?

BTW, Base now has counting `sort` but not `sortperm`, right?

Reply:
Does Base indeed have counting sort? I could not find it. I thought it only had radix sort.

I have added the comments how these functions work.

Reply:
It's here:
https://github.com/JuliaLang/julia/blob/12e679cabbe827d3be1869b9eaac24263415ee95/base/sort.jl#L842
But AFAICT that's not implemented for `sortperm`.

Reply:
Ah - that PR was made after I made my PR. Still - as you say - we need `sortperm` + we need backward compatibility.